group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,413 @@
+        currentAssignment.forEach((topicId, partitions) -> {
+            if (subscriptionIds.contains(topicId)) {
+                partitions.forEach(partition -> {
+                    TopicIdPartition topicIdPartition = new TopicIdPartition(topicId, partition);
+                    if (rackInfo.useRackStrategy && rackInfo.racksMismatch(memberId, topicIdPartition)) {
+                        currentPartitionOwners.put(topicIdPartition, memberId);
+                    } else {
+                        validCurrentAssignmentList.add(topicIdPartition);
+                    }
+                });
+            } else {
+                LOG.debug("The topic " + topicId + " is no longer present in the subscribed topics list");
+            }
+        });
+
+        return validCurrentAssignmentList;
+    }
+
+    /**
+     * Allocates the unassigned partitions to unfilled members present in the same rack in a round-robin fashion.
+     */
+    private void rackAwareRoundRobinAssignment() {
+        Queue<String> roundRobinMembers = new LinkedList<>(unfilledMembers.keySet());
+
+        // Sort partitions in ascending order by number of potential members with matching racks.
+        // Partitions with no potential members in the same rack aren't included in this list.
+        List<TopicIdPartition> sortedPartitions = rackInfo.sortPartitionsByRackMembers(unassignedPartitions);
+
+        sortedPartitions.forEach(partition -> {
+            boolean assigned = false;
+            for (int i = 0; i < roundRobinMembers.size() && !assigned; i++) {
+                String memberId = roundRobinMembers.poll();
+                Integer remainingPartitionCount = unfilledMembers.getOrDefault(memberId, 0);
+
+                if (remainingPartitionCount > 0 && !rackInfo.racksMismatch(memberId, partition)) {
+                    assignPartitionToMember(memberId, partition);
+                    assigned = true;
+                    unassignedPartitions.remove(partition);
+                }
+
+                // Only re-add the member to the end of the queue if it's still available for assignment.
+                if (unfilledMembers.containsKey(memberId)) {
+                    roundRobinMembers.add(memberId);
+                }
+            }
+        });
+    }
+
+    /**
+     * Allocates the unassigned partitions to unfilled members in a round-robin fashion.
+     *
+     * If the rack-aware strategy is enabled, partitions are attempted to be assigned back to their current owners first.
+     * This is because pure stickiness without rack matching is not considered initially.
+     *
+     * If a partition couldn't be assigned to its current owner due to the quotas OR
+     * if the rack-aware strategy is not enabled, the partitions are allocated to the unfilled members.
+     */
+    private void unassignedPartitionsRoundRobinAssignment() {
+        Queue<String> roundRobinMembers = new LinkedList<>(unfilledMembers.keySet());
```

Comment:
rackAware round robin might not be called in every case, it is only called when useRackAware strategy is true. I could make it a global attribute maybe but that seems odd since the queue is only used in these two methods.

Reply:
I changed it now so we don't need the queue at all